### ISSUE TITLE
ReadRange in ChainReplication can read from incorrect log units.

### ISSUE THREAD

**Issue by zalokhan** [CorfuDB/CorfuDB#1795]
## Overview

https://github.com/CorfuDB/CorfuDB/blob/2ab39da845ba3e40a66e31ad550d54d59e3db103/runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java#L109-L118

From the above code we can see that the chain replication protocol gets the logunit client based on only the start address.
This is incorrect as the start address can lie in a different stripe or even a different segment.
Since different log units would be responsible for serving reads in different chains, this read-range would lead to an incorrect read such as missing data.

**Comment by zalokhan** [comment]
Test proving this:

```

    @Test
    public void xyz() {

        addServer(SERVERS.PORT_0);
        addServer(SERVERS.PORT_1);
        addServer(SERVERS.PORT_2);

        TestRule blockAll = new TestRule().drop().always();

        Layout l = new TestLayoutBuilder()
                .addLayoutServer(SERVERS.PORT_0)
                .addLayoutServer(SERVERS.PORT_1)
                .addLayoutServer(SERVERS.PORT_2)
                .addSequencer(SERVERS.PORT_0)
                .addSequencer(SERVERS.PORT_1)
                .addSequencer(SERVERS.PORT_2)
                .buildSegment()
                .setStart(0L)
                .setEnd(1L)
                .setReplicationMode(Layout.ReplicationMode.CHAIN_REPLICATION)
                .buildStripe()
                .addLogUnit(SERVERS.PORT_0)
                .addToSegment()
                .addToLayout()
                .buildSegment()
                .setStart(1L)
                .setEnd(-1L)
                .setReplicationMode(Layout.ReplicationMode.CHAIN_REPLICATION)
                .buildStripe()
                .addLogUnit(SERVERS.PORT_0)
                .addLogUnit(SERVERS.PORT_1)
                .addToSegment()
                .addToLayout()
                .build();
        bootstrapAllServers(l);
        addClientRule(getManagementServer(SERVERS.PORT_0).getManagementAgent().getCorfuRuntime(), blockAll);
        addClientRule(getManagementServer(SERVERS.PORT_1).getManagementAgent().getCorfuRuntime(), blockAll);
        addClientRule(getManagementServer(SERVERS.PORT_2).getManagementAgent().getCorfuRuntime(), blockAll);

        CorfuRuntime corfuRuntime = CorfuRuntime.fromParameters(CorfuRuntimeParameters.builder()
                .layoutServer(NodeLocator.parseString(SERVERS.ENDPOINT_0))
                .build()).connect();

        byte[] payload = "test".getBytes();
        IStreamView streamView = corfuRuntime.getStreamsView().get(CorfuRuntime.getStreamID("testStream"));

        addClientRule(corfuRuntime, SERVERS.ENDPOINT_1, new TestRule().drop().always());

        streamView.append(payload);
        CompletableFuture.supplyAsync(() -> streamView.append(payload));

        Sleep.SECONDS.sleepUninterruptibly(5);

        ChainReplicationProtocol chainReplicationProtocol = new ChainReplicationProtocol(new ReadWaitHoleFillPolicy(Duration.ofSeconds(1), Duration.ofMillis(100)));
        Map<Long, ILogData> map = chainReplicationProtocol.readRange(corfuRuntime.getLayoutView().getRuntimeLayout(), Range.closed(0L, 1L), false);
        assertThat(map.get(0L).getType()).isEqualByComparingTo(DataType.DATA);
        assertThat(map.get(1L).getType()).isEqualByComparingTo(DataType.DATA);

        clearClientRules(getManagementServer(SERVERS.PORT_0).getManagementAgent().getCorfuRuntime());
        clearClientRules(getManagementServer(SERVERS.PORT_1).getManagementAgent().getCorfuRuntime());
        clearClientRules(getManagementServer(SERVERS.PORT_2).getManagementAgent().getCorfuRuntime());

        while (corfuRuntime.getLayoutView().getLayout().getSegments().size() != 1) {
            corfuRuntime.invalidateLayout();
            Sleep.sleepUninterruptibly(PARAMETERS.TIMEOUT_VERY_SHORT);
        }
        addServerRule(SERVERS.PORT_0, new TestRule().always().drop());
        Sleep.SECONDS.sleepUninterruptibly(5);
        clearClientRules(corfuRuntime);
        corfuRuntime.invalidateLayout();

        map = chainReplicationProtocol.readRange(corfuRuntime.getLayoutView().getRuntimeLayout(), Range.closed(0L, 1L), false);
        assertThat(map.get(0L).getType()).isEqualByComparingTo(DataType.DATA);
        assertThat(map.get(1L).getType()).isEqualByComparingTo(DataType.DATA);
    }
```

**Comment by zalokhan** [comment]
Start 3 servers and 2 segments [0, 1) and [1, infinity) in the layout.
Disable all management agents so that the segments do not get merged.
We write to 2 addresses:
Address 0 is written only to SERVERS_0
Address 1 is partially written. It's written to SERVERS_0 but not replicated to SERVERS_1.

We now do a range read. The data is read:
for address 0 it's read from SERVERS_0 and returns a valid payload.
for address 1 it's read from SERVERS_0 and returns a valid payload. Although the value obtained is correct, the node from which this is read is incorrect. Later in the test, we see how this can cause inconsistency.

We now enable the management agents and allow the segments to be merged.
Note that the payload at address 1 is still not fully replicated and still only exists on SERVERS_0.
We now disable SERVERS_0 so that it is removed from the chain and marked unresponsive.

Finally, we attempt to do a range read again.
As expected, we get,
for address 0 it's read from SERVERS_1 and returns a valid payload.
for address 1 it's read from SERVERS_1 and returns empty. (Inconsistent with the previous read)

**Comment by WenbinZhu** [comment]
Addressed in #1855

**Comment by WenbinZhu** [comment]
Addressed in #1855
